### PR TITLE
Use TypeMeta::dtor() instead of Blob::DestroyCall

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #12307 Add TypeMeta::New/Delete&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D10184116/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#11500 Use TypeMeta::dtor() instead of Blob::DestroyCall**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9763423/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11925 Serialization takes pointers instead of Blob&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9884177/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #11926 Remove Blob::ShareExternal from serialization&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D9884563/)

Since TypeMeta already stores a destructor, and we removed the ability from Blob to store a custom destructor in a diff stacked below this, there is now no reason for Blob to store it again.

Differential Revision: [D9763423](https://our.internmc.facebook.com/intern/diff/D9763423/)

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: caffe2/core/blob.h [hunk 1]
```diff
@@ -241,16 +233,16 @@ class CAFFE2_API Blob final : public c10::intrusive_ptr_target {
...
-  /**
-   * @brief A destroy call that is used to properly deconstruct objects.
-   */
-  template <class T>
-  static void Destroy(void* pointer) {
-    delete static_cast<T*>(pointer);
+  void free_() {
+   if (has_ownership_) {
+     CAFFE_ENFORCE(pointer_ != nullptr, "Can't have ownership of nullptr");
+     (*meta_.dtor())(pointer_, 1);
```
Comment (dzhulgakov) [inline caffe2/core/blob.h]: wait a second - before the delete call was actually freeing the memory (calling operator `delete`), now it just calls placement destructor and leaves memory hanging around

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

**Review by ezyang — CHANGES_REQUESTED** [review]
Agreed with Dmytro

File: caffe2/core/blob.h (hunk 1 above)
Comment (smessmer) [reply caffe2/core/blob.h]: fixed

**Comment by smessmer** [comment]
fixed

**Review by ezyang — APPROVED** [review]
how fancy

**Review by Yangqing — APPROVED** [review]

**Pull request closed without merging**